Azure/azure-cli
# {Package} Migrate CLI packaging system to PEP420

**Description**  

Background
----------
[PEP420 ](https://www.python.org/dev/peps/pep-0420/)is an official python enhancement proposal which introduces implicit namespace package. 

There are 3 ways to declare a namespace package.

- pkg_resource style
- pkgutil style
- native style (PEP420)

Currently, azure cli is using pkg_resource style namespace package, which is no longer recommended per python official document. Besides, pkg_resource style namespace package performs slow when importing python modules.

This PR is to migrate azure cli namespace package to native style. Refer [Packaging namespace packages ](https://packaging.python.org/guides/packaging-namespace-packages/)to see details about above 3 namespace package styles.

There is also a guide about how to do azure sdk package. Refer [Azure packaging ](https://github.com/Azure/azure-sdk-for-python/blob/master/doc/dev/packaging.md)to see the details. 

We will follow this guide to to do azure cli packaging.

Goal
----------
The main goal of this task to to remove `__init__.py` file under `azure` and `cli` folder in our final distributions on different platforms, includes:

- MSI
- DEB
- RPM
- Pypi
- Homebrew

Changes
----------
In order to achieve this, several changes are needed as below:

**azure-cli-dev-tools**
- remove installation of all cli related `-nspkg` packages([PR](https://github.com/Azure/azure-cli-dev-tools/pull/240))

**some python sdk dependency**
- azure-cosmos
- azure-multiapi-storage

**azure-cli**
- remove all `-nspkg` dependency
- remove `azure-cli-nspkg` and `azure-cli-command_modules-nspkg`
- update `__init__.py`, `setup.py` and `MANIFEST.in` according to azure packaging guide
- move `__version__` and `__author__` to `__main__.py`
- update `azure_bdist_wheel` to fix package `__main__.py` problem
- sdk dependency
- fix test cases

Plan
----------

1. merge `azure-cli-dev-tools` [PR](https://github.com/Azure/azure-cli-dev-tools/pull/240)
2. build new `azdev` version
3. rollback `azdev` updates in this PR
4. merge PR


**Testing Guide**  
- Install `azure-cli` on different platform, and see whether `__init__.py` still exists
- Make sure all previous tests are passed in both current CI pipelines and Feng's new CI pipelines.
- Make sure `.tar.gz` and `.whl` files are same with previous ones.

**History Notes**  


[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.  
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.
